pkg/controller/common.go

```diff
@@ -597,335 +600,805 @@ func (c *SharedState) isVolumeVisited(azVolumeName string) bool {
+	for _, node := range nodes {
+		candidateNodes.add(node.Name)
+	}
 
-		if podObj.Spec.Affinity == nil || podObj.Spec.Affinity.PodAffinity == nil {
+	for _, pod := range p.pods {
+		if pod.Spec.Affinity == nil || pod.Spec.Affinity.PodAntiAffinity == nil {
 			continue
 		}
-		// TODO: add filtering for pod anti-affinity and more advanced affinity scenarios (e.g. Preferred affinites)
-		for _, podAffinity := range podObj.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
-			podSelector, err := metav1.LabelSelectorAsSelector(podAffinity.LabelSelector)
+
+		for _, podAntiAffinity := range pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			podSelector, err := metav1.LabelSelectorAsSelector(podAntiAffinity.LabelSelector)
 			// if failed to convert pod affinity label selector to selector, log error and skip
 			if err != nil {
-				klog.Errorf("failed to convert pod affinity (%v) to selector", podAffinity.LabelSelector)
+				klog.Errorf("failed to convert pod anti-affinity (%v) to selector", podAntiAffinity.LabelSelector)
 				continue
 			}
 			requirements, selectable := podSelector.Requirements()
 			if selectable {
-				podAffinities = podAffinities.Add(requirements...)
+				podAntiAffinities := labels.NewSelector().Add(requirements...)
+				pods := []v1.Pod{}
+
+				if len(podAntiAffinity.Namespaces) > 0 {
+					for _, namespace := range podAntiAffinity.Namespaces {
+						podList := &v1.PodList{}
+						if err = p.azr.getClient().List(ctx, podList, &client.ListOptions{LabelSelector: podAntiAffinities, Namespace: namespace}); err != nil {
+							klog.Errorf("failed to retrieve pod list: %v", err)
+							continue
+						}
+						pods = append(pods, podList.Items...)
+					}
+				} else {
+					podList := &v1.PodList{}
+					if err = p.azr.getClient().List(ctx, podList, &client.ListOptions{LabelSelector: podAntiAffinities}); err != nil {
+						klog.Errorf("failed to retrieve pod list: %v", err)
+						continue
+					}
+					pods = podList.Items
+				}
+
+				nodesWithSelectedPods := set{}
+				for _, pod := range pods {
+					nodesWithSelectedPods.add(pod.Spec.NodeName)
+				}
+
+				qualifiedNodes := set{}
+				for node := range nodesWithSelectedPods {
+					var nodeInstance v1.Node
+					if err = p.azr.getClient().Get(ctx, types.NamespacedName{Name: node.(string)}, &nodeInstance); err != nil {
+						klog.Errorf("failed to get node (%s)", node.(string))
+						continue
+					}
+					if value, exists := nodeInstance.Labels[podAntiAffinity.TopologyKey]; exists {
+						// get list of node candidates
+						nodeRequirement, err := azureutils.CreateLabelRequirements(podAntiAffinity.TopologyKey, selection.NotEquals, value)
+						if err != nil {
+							continue
+						}
+						nodeSelector := labels.NewSelector().Add(*nodeRequirement)
+
+						nodeList := &v1.NodeList{}
+						if err = p.azr.getClient().List(ctx, nodeList, &client.ListOptions{LabelSelector: nodeSelector}); err != nil {
```

Comment:
Since we are using qualifiedNodes only to filter out the candidate nodes, can't we filter out the candidate nodes directly here by looping over the candidateNodes and matching the nodeSelectors to each of the candidate node. With that, we can avoid making a List call here.